web/js/codeworld_shared.js

```diff
@@ -132,95 +132,157 @@ var hintBlacklist = [
     "pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+// codeWorldSymbols is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldSymbols = {
+//   codeWorldLogo: {
+//     declaration: "codeWorldLogo :: Picture",
+//     symbolStart: 0,
+//     symbolEnd: 13,
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldSymbols = {},
+    codeWorldBuiltinSymbols = {};
+
+function getWordStart(word, line) {
+    return line.indexOf(word);
+}
 
-function onHover(cm, data, node){
-    if (data && data.token && data.token.string) {
-        var token_name = data.token.string;
-        if (hintBlacklist.indexOf(token_name) == -1 && codeWorldBuiltinsDocs[token_name]){
-            return codeWorldBuiltinsDocs[token_name];
-        }
+function getWordEnd(word, line) {
+    var wordStart = getWordStart(word, line);
+    if (wordStart != -1) {
+        return wordStart + word.length
     }
-    return;
+    return -1;
 }
 
-// Hints and hover tooltips
-function registerStandardHints(successFunc)
-{
-    function createHint(line, wordStart, wordEnd, cname) {
-        var word = line.slice(wordStart, wordEnd);
-        if (!cname) cname = 'hint-word';
+function parseSymbolsFromCurrentCode() {
+    var lines = window.codeworldEditor.getValue().split('\n'),
+        word = null,
+        parseResults = {};
 
-        function renderer(elem, data, cur) {
-            if (wordStart > 0) {
-                elem.appendChild(document.createTextNode(line.slice(0, wordStart)));
+    lines.forEach(function(line) {
+        // f(x, y) =
+        if (/^\w+\(.*/.test(line)) {
+            word = line.split("(")[0].trim();
+            parseResults[word] = {};
+        }
+        // foo =
+        else if (/^\S+\s*=/.test(line)) {
```

Comment:
There are definitely weird cases that will accidentally match here.  For instance:

    {- ==================
    Some block comment
    ================== -}

The first line will match and add a hint for "{-"

Also:

    a*&*b = a + 2 * b

This is actually a definition of the `*&*` operator, but will be misparsed as a definition for the symbol `a*&*b`.

I think this is okay for the initial merge, but should definitely file a bug (probably the same one as in the earlier comment) to replace this with an honest-to-goodness parser.  At the same time, we should make it able to recognize definitions that are split across line breaks, like

    f(x, y)
      = x + y

So let's leave this undone for now, but make sure to track the future work.

Reply:
Will add this to #798